src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+ */
+
+int SUNAdaptController_SetParams_ExpGus(SUNAdaptController C,
+                                        sunrealtype k1, sunrealtype k2)
+{
+  SODERLIND_K1(C) = k1+k2;
+  SODERLIND_K2(C) = -k2;
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an implicit Gustafsson controller
+ */
+
+SUNAdaptController SUNAdaptController_ImpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_ImpGus(C,
+                                             DEFAULT_IMPGUS_K1,
+                                             DEFAULT_IMPGUS_K2);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set explicit Gustafsson parameters
+ */
+
+int SUNAdaptController_SetParams_ImpGus(SUNAdaptController C,
+                                        sunrealtype k1, sunrealtype k2)
+{
+  SODERLIND_K1(C) = k1+k2;
+  SODERLIND_K2(C) = -k2;
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(1.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+
+
+/* -----------------------------------------------------------------
+ * implementation of controller operations
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController_Type SUNAdaptController_GetType_Soderlind(SUNAdaptController C)
+{ return SUN_ADAPTCONTROLLER_H; }
+
+int SUNAdaptController_EstimateStep_Soderlind(SUNAdaptController C, sunrealtype h,
+                                              int p, sunrealtype dsm, sunrealtype* hnew)
+{
+  /* order parameter to use */
+  const int ord = p + 1;
+
+  /* set usable time-step adaptivity parameters */
+  const sunrealtype k1 = -SODERLIND_K1(C) / ord;
+  const sunrealtype k2 = -SODERLIND_K2(C) / ord;
+  const sunrealtype k3 = -SODERLIND_K3(C) / ord;
+  const sunrealtype k4 = SODERLIND_K4(C);
+  const sunrealtype k5 = SODERLIND_K5(C);
+  const sunrealtype e1 = SUNMAX(SODERLIND_BIAS(C) * dsm, TINY);
+  const sunrealtype e2 = SUNMAX(SODERLIND_EP(C), TINY);
+  const sunrealtype e3 = SUNMAX(SODERLIND_EPP(C), TINY);
+  const sunrealtype hrat = h / SODERLIND_HP(C);
+  const sunrealtype hrat2 = SODERLIND_HP(C) / SODERLIND_HPP(C);
+
+  /* compute estimated optimal time step size */
+  if (SODERLIND_FIRSTSTEPS(C) < 1) {
+    *hnew = h * SUNRpowerR(e1,k1);
+  } else if  (SODERLIND_FIRSTSTEPS(C) < 2) {
+    *hnew = h * SUNRpowerR(e1,k1) * SUNRpowerR(e2,k2) * SUNRpowerR(hrat,k4);
+  } else {
+    *hnew = h * SUNRpowerR(e1,k1) * SUNRpowerR(e2,k2) * SUNRpowerR(e3,k3)
+              * SUNRpowerR(hrat,k4) * SUNRpowerR(hrat2,k5);
+  }
+
+  /* return with success */
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_Reset_Soderlind(SUNAdaptController C)
+{
```

Comment:
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

Reply:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.